Adding simple Get call on client, to handle boilerplate

StabbyCutyou opened the pull request:
Tossing this as an idea out there - A simple series of Get / Set / Delete methods to make the really common cases a little less boiler-platey to work with.

Example of how I'd see someone using this:

```golang
objs, err := client.Get(&GetOptions{
  Bucket: "b",
  Key :"y",
  BucketType: "bt"
})

if err != nil {
  // stuff
}
```

The idea being, only one error to check, optional BucketType. I think you could do something similar for Set / Delete, would be happy to implement those if you think this approach makes sense.

lukebakken commented:
Hello!

Thank you for taking the time to open this PR. This sort of thing was indeed on my radar to do once someone asked for it. I will review the PR with some suggestions. Implementing Fetch, Store and Delete would be the logical starting place.

lukebakken commented on client.go:
```diff
@@ -118,3 +124,30 @@ func newClientUsingAddresses(port uint16, remoteAddresses []string) (*Client, er
 }
+
+// Get is a simple invocation of the NewFetchValueCommand
+func (c *Client) Get(opts *GetOptions) ([]*Object, error) {
```
To be consistent this should be named `Fetch` rather than `Get`. I'm not sure if an entirely separate options struct is necessary as there are only three options that are also the same as will be used by `Delete`.

StabbyCutyou replied:
Good points on both.

The only reason I initially broke the "Options" out was so for Set, you can have some flexibility for how the client can set any 2i properties, but maybe if they need that, they need something above a simple Set?

Either way, happy to make these changes

lukebakken commented on client.go:
```diff
@@ -118,3 +124,30 @@ func newClientUsingAddresses(port uint16, remoteAddresses []string) (*Client, er
+		WithBucket(opts.Bucket).
+		WithKey(opts.Key)
+
+	if opts.BucketType != "" {
```
There's no need for this check, `.WithBucketType` accepts a 0-length string and [the validation here](https://github.com/basho/riak-go-client/blob/master/locatable.go#L28-L30) will ensure the default bucket type is used in this case.

StabbyCutyou replied:
Ah, good to know!

lukebakken commented:
An integration test in `client_i_test.go` would be great to demonstrate Store, Fetch and Delete using these methods.

StabbyCutyou commented:
If you feel this approach has merit, and is likely to be accepted (with Set and Delete), then I'll gladly spend the time to write integration tests for it.

Didn't wanna do that up front incase this was not inline with the vision for the lib.

lukebakken commented:
I figure if someone wants a feature enough to implement it and provide tests, I'll always accept it :smile:

StabbyCutyou commented:
Deal - I'll work on this in my spare time over the next 2-3 days, and ping
you when the changes are ready.

On Tue, Nov 10, 2015 at 11:22 AM, Luke Bakken <notifications@github.com>
wrote:

> I figure if someone wants a feature enough to implement it and provide
> tests, I'll always accept it [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/basho/riak-go-client/pull/37#issuecomment-155471761>.
>

christophermancini commented:
:+1: what @lukebakken said! :)

StabbyCutyou commented:
I updated the PR and renamed it Fetch, to be consistent, dropped the Optionsstruct, and added methods for Store and Delete. Before I add tests, want to make sure you're happy with the proposed API

StabbyCutyou commented on client.go:
```diff
@@ -118,3 +118,80 @@ func newClientUsingAddresses(port uint16, remoteAddresses []string) (*Client, er
+		Charset:         "utf-8",
+		ContentEncoding: "utf-8",
+		Value:           data,
+	}
```
Not sure how you best want to handle letting the user configure this... perhaps delegate the object creation to the calling code, and have this method just accept a *riak.Object instead of an array of bytes?

lukebakken replied:
Yes, the last parameter should be `*Object` like you suggest.

lukebakken commented on client.go:
```diff
@@ -118,3 +118,73 @@ func newClientUsingAddresses(port uint16, remoteAddresses []string) (*Client, er
+		WithBucketType(bucketType).
+		WithBucket(bucket).
+		WithKey(key).
+		WithContent(obj).
```
Add `.WithReturnBody(true)` to the builder so the object is returned from Riak.

lukebakken commented on client.go:
```diff
@@ -118,3 +118,73 @@ func newClientUsingAddresses(port uint16, remoteAddresses []string) (*Client, er
+}
+
+// Delete is a simple invocation of the NewFetchValueCommand
+func (c *Client) Delete(bucketType, bucket, key string) error {
```
Add a `vclock []byte` argument. It doesn't *have* to be set but if your bucket has `allow_mult` set to true (default for 2.0 buckets) you'll need a vclock to prevent siblings.

lukebakken commented:
Looks good, just a few suggestions. Test away :smile:

StabbyCutyou commented:
FYI I've been stalled on this due to some time-consuming issues in my personal life (new house, have to rebuild the bathroom), but I'm still working on it.

jaytaylor commented:
+1 for getting this merged!

lukebakken commented:
Closing due to inactivity.
